### PR TITLE
Block the request to issue new tokens if an invalid provider is given.

### PR DESCRIPTION
**Pull request by Alymosul**
This PR adds an early protection in the AddCustomProvider middleware by checking that the provider given in the request is registered in the auth configuration, if not it will throw an exception to block the request also if the request is missing the provider param it will block it.

### CODE REVIEW

File: src/Http/Middleware/AddCustomProvider.php [hunk 1]
```diff
@@ -51,4 +56,26 @@ public function terminate()
...
+            return true;
+        }
+
+        foreach (config('auth.guards') as $guardsConfiguration) {
+            if ($guardsConfiguration['provider'] === $provider) {
+                return false;
+            }
+        }
+
+        return false;
```
Comment (sfelix-martins) [inline src/Http/Middleware/AddCustomProvider.php]: @Alymosul Has an error here. If you pass a provider that not found on `config('auth.guards')` the `OAuthServerException` should be thrown. But returning false the provider don't will be invalid.

File: tests/Unit/AddCustomProviderTest.php [hunk 2]
```diff
@@ -25,9 +25,7 @@ public function tearDown()
     public function testIfApiProviderOnAuthWasSetCorrectly()
```
Comment (sfelix-martins) [inline tests/Unit/AddCustomProviderTest.php:26]: If you change the return of the method `invalidRequest($provider)` to `true` this test will fail. Because the provider `companies` isn't defined. To it works you should set a config like `config(['auth.guards.company.provider' => 'companies']);` before execute the `AddCustomProvider::handle()` method.

File: tests/Unit/AddCustomProviderTest.php [hunk 3]
```diff
@@ -25,9 +25,7 @@ public function tearDown()
...
     {
         $request = Mockery::mock(Request::class);
-        $request->shouldReceive('all')->andReturn([
-            'provider' => 'companies',
-        ]);
+        $request->shouldReceive('get')->andReturn('companies')->with('provider');
 
         $middleware = new AddCustomProvider();
         $middleware->handle($request, function () {
```
Comment (sfelix-martins) [inline tests/Unit/AddCustomProviderTest.php:35]: To improve the test coverage you can add tests simulating pass a `null` provider and pass a provider not existent on `config('auth.guards')`. E.g.:

```php
    public function testPassNotExistentProvider()
    {
        $this->expectException(OAuthServerException::class);

        $request = Mockery::mock(Request::class);
        $request->shouldReceive('get')->andReturn('not_found')->with('provider');

        $middleware = new AddCustomProvider();
        $middleware->handle($request, function () {
            return 'response';
        });
    }

    public function testDoNotPassProviderToRequest()
    {
        $this->expectException(OAuthServerException::class);

        $request = Mockery::mock(Request::class);
        $request->shouldReceive('get')->andReturn(null)->with('provider');

        $middleware = new AddCustomProvider();
        $middleware->handle($request, function () {
            return 'response';
        });
    }
```

**Review by sfelix-martins — CHANGES_REQUESTED** [review]
@Alymosul Firstly, thanks for your contribution. Just check the review comments and update to able the merge.

File: src/Http/Middleware/AddCustomProvider.php [hunk 4]
```diff
@@ -51,4 +56,26 @@ public function terminate()
...
+     * Check if the given provider is not registered in the auth configuration file.
+     *
+     * @param $provider
+     *
+     * @return bool
+     */
+    protected function invalidProvider($provider)
+    {
+        if (is_null($provider)) {
+            return true;
```
Comment (sfelix-martins) [inline src/Http/Middleware/AddCustomProvider.php:70]: I've been thinking, if a null provider is invalid we will broke the compatibility, because actually, if none provider was passed for the request the default provider from `api` guard , generally `users`, is used. The plugin users will have to change the request to `oauth/token` passing the default provider on `api` guard to `provider` request param.

What you think about this? Broke the compatibility and up to `v3.0`? Or do we leave it as it is?

File: tests/Unit/AddCustomProviderTest.php (hunk 3 above)
Comment (Alymosul) [reply tests/Unit/AddCustomProviderTest.php:35]: We'll add them.

File: src/Http/Middleware/AddCustomProvider.php (hunk 1 above)
Comment (Alymosul) [reply src/Http/Middleware/AddCustomProvider.php]: Yeah! I missed that, i'll fix it.

File: src/Http/Middleware/AddCustomProvider.php (hunk 4 above)
Comment (Alymosul) [reply src/Http/Middleware/AddCustomProvider.php:70]: To me, i think you should never guide an authentication request to pass...
The request must contain all the correct params and if any of them is invalid or missing we should stop the request from propagating to the next layers of the app.

File: tests/Unit/AddCustomProviderTest.php (hunk 2 above)
Comment (sfelix-martins) [reply tests/Unit/AddCustomProviderTest.php:26]: Just change the method `AddCustomProvider::invalidRequest` to return `true` after check if provider exists. When you run the tests this test should fails. Because the provider `companies` doesn't exists on config. Do you understood?

Comment (Alymosul) [reply tests/Unit/AddCustomProviderTest.php:26]: Yeah! going to fix that.

File: src/Http/Middleware/AddCustomProvider.php (hunk 4 above)
Comment (sfelix-martins) [reply src/Http/Middleware/AddCustomProvider.php:70]: Makes sense. Please just change the branch target from `2.0` to new branch `3.0`. As this change will break the compatibility we will have to create a major upgrade in version. 

[Follow this guide](https://blog.github.com/2016-08-15-change-the-base-branch-of-a-pull-request/) to change the target.

What you think about change the php required version to `"php": ">=7.1"`?

Comment (Alymosul) [reply src/Http/Middleware/AddCustomProvider.php:70]: I think there's no need to change it.

File: tests/Unit/AddCustomProviderTest.php (hunk 3 above)
Comment (sfelix-martins) [reply tests/Unit/AddCustomProviderTest.php:35]: Good. Thanks.

**Review by sfelix-martins — APPROVED** [review]
